src/sync.cpp

```diff
@@ -123,8 +123,12 @@ static void push_lock(void* c, const CLockLocation& locklocation)
     }
 }
 
-static void pop_lock()
+static void pop_lock(void* cs)
 {
+    // We assert here that locks are popped in the order they were locked.
+    // This is a super-overly-restrictive requirement, but we need it to
+    // make our deadlock detection work properly.
+    assert((*lockstack).back().first == cs);
```

Comment:
If I'm understanding this correctly, I think it may cause issues with recursive locks that are currently harmless (though potentially troublesome)? e.g.:
```
recursive_mutex a;
recursive_mutex b;
lock(a)
lock(a)
lock(b)
lock(b)
unlock(a)
unlock(b)
unlock(b)
unlock(a)
```

Reply:
Yes, those will now be DEBUG_LOCKORDER failures, but it appears to currently not be violated, and would detect some potentially significant issues, consider:

lock(a)
lock(b)
unlock(a) (this now pops b off the stack, so lockorder thinks we have a locked)
lock(a) (this is a deadlock cause we've now violated lockorder, but DEBUG_LOCKORDER thinks that we're just re-locking a).